doc/engine.md

```diff
@@ -0,0 +1,215 @@
+code for the client to execute, so all game code
+is effectively executed only on the server.
+
+The client only receives media assets and instructions on how
+to render them, and only handles a rudimentary portion
+of the gameplay which would be too slow to control remotely,
+such as player movement and collision.
+
+This actually allows clients to connect to servers to
+discover and play games and mods without the need to
+install them ahead of time, which is a unique and
+valuable feature of the Minetest ecosystem,
+differentiating it from other voxel game engines and clones.
+
+Therefore, any API exposed to modders and game developers
+is implemented in the form of server-side functionality,
+which may then trigger action on the client environment by
+means of fixed-function remote network commands.
+
+This has some consequences for the design of engine APIs
+and features. Note that introduction of server-sent scripts
+in the future may relax and change some rules, but this
+is what we're stuck with for now.
+
+First of all, every feature that requires
+support in the client environment adds to the amount of
+individual network packets and code to support them,
+as well as client code which implements the functionality.
+
+Therefore, features that require communicating with the client
+shall be designed to maximize the functionality gained from them,
+and to support as many possible use cases as possible.
+
+Second, game code is not able to control client side
+capabilities without explicit API support for it.
+The consequences of this are expanded upon in the following sections.
+
+Lastly, even for features that do not require communication
+with the client, consider the maintenance cost, and whether it is
+possible to achieve the same goal with reasonable performance
+using Lua scripts alone - if it is, it probably does not need
+an engine feature. Resist the temptation to add helper functions
+for every single thing, mods and libraries can do it just fine.
+
+### 1.1. PR design checklist
+
+Here is a list of questions you can ask yourself that can help you
+triage whether or not an engine change is necessary or useful.
+
+* Am I solving an actual problem?
+* Am I just dealing with a legitimately hard computer science problem, and not a limitation?
+* Can the problem be framed in a more general way?
+* Is it possible to do what I want with Lua scripts?
+* Is the performance of the Lua solution acceptable?
+* Is this gonna be useful to others as well?
+* Is the solution robust and not just a hack?
+
+## 2. Obligation to content creators
+
+Game developers and modders should be provided with a
+reasonable, stable working environment to create their content in.
+
+At no point should you pull the rug from under a developer by
+changing this environment in an incompatible way.
+
+New features should primarily serve the purpose of providing
+content creators with more freedom. Features that restrict creative
+freedom in any way are unacceptable.
```

Comment:
I avoided providing examples in the whole document because the respective authors could get offended.

In principle, visual PRs with no opt-outs qualify as restrictive by knocking out possible art styles.

Surely you could find at least one example of a gameplay restrictive PR or request if you looked hard enough. Of course it would get closed pretty quickly.

I agree that we get a lot more bloaty and shortsighted PRs and requests than outright restrictive ones.